clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -440,6 +440,7 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
             correlate(req.request().header(), header);
             if (!metadataUpdater.maybeHandleCompletedReceive(req, now, body))
                 responses.add(new ClientResponse(req, now, false, body));
+            inFlightRequests.completeNext(source);
```

Comment:
@becketqin : Yes, if it's a bug, it's going to be hard to auto fixing it in the client code. Just propagating the exception to the caller is probably the best that we can do.